library/core/src/iter/adapters/step_by.rs

```diff
@@ -32,12 +36,12 @@ where
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        if self.first_take {
-            self.first_take = false;
-            self.iter.next()
-        } else {
-            self.iter.nth(self.step)
-        }
+        // branchless step size selection.
```

Comment:
Is branchless really better than letting the branch predictor run here?

Also, just using an `if` will give you a `select` in LLVM: <https://rust.godbolt.org/z/vGcKjrdc3>

I think the only reason you're getting something different from the `if` with this is that LLVM isn't smart enough to realize that you wrote a `select` here, because it ignores us telling it that the `load` of the bool will always be zero or one.

So I'm inclined to say that you also found a bug here worth filing on LLVM.

I'm not convinced that the mask hack is worth it here, though.

Can it not just be the simpler
```
let step_size = if self.first_take { 0 } else { self.step };
self.first_take = false;
self.iter.nth(step_size);
```
to get all the "only one instantiation" goodness, but still probably let LLVM understand what's going on and lower the `select` as it thinks is best?

Reply:
Fun....

```
branchless:
    iter::bench_range_step_by       38.00ns/iter +/- 0.00ns
if-else:
    iter::bench_range_step_by       69.00ns/iter +/- 1.00ns
```

But that's with `-Ctarget-cpu=native`!
(and idk why it got even faster than what's in the PR description)

With x86-64 baseline I get:

```
branchless:
    iter::bench_range_step_by       39.00ns/iter +/- 0.00ns
if-else:
    iter::bench_range_step_by       39.00ns/iter +/- 0.00ns
```

😮‍💨